### PR TITLE
NEW Add AccountResetUI, with supporting endpoint for sending reset email

### PR DESCRIPTION
**Pull request by Cheddam**
This is a fairly heavy PR - sorry! I'll be splitting the UI / underpinnings for actually performing the account reset into a separate PR.

This PR:

- Introduces the `AccountResetUI` component, which renders within the `RegisteredMFAMethodListField` component (which really needs to be renamed now) when an MFA administrator views a profile with read-only access.
- Renames `MemberExtension` to `MemberMFAExtension`, and adds `MemberResetExtension` for Account Reset-specific fields / methods
- Adds a functional `POST /admin/security/reset/$ID` endpoint for sending account reset emails, along with an email template.
- Tweaks `LoadingIndicator` to drop its built-in margins and to allow adjustment of its size via a prop
  - Other usages of `LoadingIndicator` have been wrapped in an `.mfa-loading` element which reinstates the previous margins

**TODO:**

- [x] Write tests
- [x] Add CSRF to the Account Reset endpoint
- [x] Add icons to the 'success' and 'failure' messages
- [ ] ~Consider splitting the `AccountResetUI` component into a separate FormField, otherwise rename `RegisteredMFAMethodListField` to something more descriptive~ (will do this in a later PR)
- [x] Tidy up styling

### CODE REVIEW

File: src/Extension/SecurityAdminAccountResetExtension.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            return $email->send();
+        } catch (Exception $e) {
+            Injector::inst()->get(LoggerInterface::class)->info('WARNING: Account Reset Email failed to send');
+            return false;
+        }
+    }
+
+    protected function getAccountResetLink(Member $member, string $token)
+    {
+        return null;
```
Comment (ScopeyNZ) [inline src/Extension/SecurityAdminAccountResetExtension.php]: Missing implementation?

Comment (ScopeyNZ) [reply src/Extension/SecurityAdminAccountResetExtension.php]: Nevermind, read the OP.

File: client/src/components/BasicMath/Login.js [hunk 2]
```diff
@@ -35,7 +35,7 @@ class Login extends Component {
     const { ss: { i18n } } = window;
 
     if (!numbers) {
-      return <LoadingIndicator />;
+      return <div className="mfa-loader"><LoadingIndicator /></div>;
```
[the remark this answers is not in the document]
Comment (Cheddam) [reply client/src/components/BasicMath/Login.js]: I should have made this class a little more distinct from the one in `LoadingIndicator`, which is `mfa__loader`. This class _is_ necessary, as it replaces a bit of the styling that used to be baked into `LoadingIndicator` but which I needed to remove to make it usable more broadly.

It might be best to change the class on `LoadingIndicator` instead - to `.loading-indicator`, since it's an independent component. Thoughts?

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js [hunk 3]
```diff
@@ -45,12 +46,22 @@ class RegisteredMFAMethodListField extends Component {
       fallbacks['MultiFactorAuthentication.DEFAULT']
     );
 
+    let className = 'registered-mfa-method-list-field';
```
[the remark this answers is not in the document]
Comment (Cheddam) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Right, I'd forgotten about `classnames` - Vue bakes this functionality in 😉 

I'll pull it in and adjust.

File: client/src/components/LoadingIndicator.js [hunk 4]
```diff
@@ -1,3 +1,3 @@
 import React from 'react';
 
-export default () => <div className="mfa__loader" />;
+export default ({ size = '6em' }) => <div style={{ width: size, height: size }} className="mfa__loader" />;
```
[the remark this answers is not in the document]
Comment (Cheddam) [reply client/src/components/LoadingIndicator.js]: Oops, I understand your understanding now! Okay, we can do this as props - I'll add one for the color, too.

File: src/Extension/MemberMFAExtension.php [hunk 5]
```diff
@@ -22,9 +22,9 @@
  * @property MethodInterface DefaultRegisteredMethod
  * @property string DefaultRegisteredMethodID
  * @property bool HasSkippedMFARegistration
- * @property Member|MemberExtension owner
+ * @property Member|MemberMFAExtension owner
  */
-class MemberExtension extends DataExtension implements PermissionProvider
+class MemberMFAExtension extends DataExtension implements PermissionProvider
```
[the remark this answers is not in the document]
Comment (Cheddam) [reply src/Extension/MemberMFAExtension.php]: Yeah, the rename was to differentiate it from the Account Reset-related MemberExtension I was adding but I definitely didn't hit the mark here. Perhaps the AccountReset classes should live in a grouped namespace, e.g. `SilverStripe\MFA\AccountReset\Extension\MemberExtension`, to make it clear that it's a self-contained feature that shouldn't necessarily live within MFA long-term?

For now, I can just revert the rename 👍

File: client/src/components/BasicMath/Login.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/BasicMath/Login.js]: You could also make the class name extensible via a passed prop

[the remark this answers is not in the document]
Comment (ScopeyNZ) [reply client/src/components/BasicMath/Login.js]: Ok I think we're onto something in the discussion below: https://github.com/silverstripe/silverstripe-mfa/pull/127#discussion_r286701883

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    this.state = {
+      complete: false,
+      failed: false,
+      submitting: false,
+    };
+
+    this.onSendReset = this.onSendReset.bind(this);
+  }
+
+  onSendReset() {
```
[the remark this answers is not in the document]
Comment (Cheddam) [reply client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js]: We should document this!

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.scss [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    color: $color-blue;
+  }
+
+  .account-reset-action--failure & {
+    color: $color-grey;
+  }
+}
+
+.account-reset-action__message {
+  margin-left: 1em;
```
[the remark this answers is not in the document]
Comment (Cheddam) [reply client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.scss]: This is now a unique spacing variable, since no existing vars matched.

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js]: This would probably be the place: https://docs.silverstripe.org/en/4/contributing/javascript_coding_conventions/

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+          }
+        </p>
+
+        { !submitting && !complete && this.renderAction() }
+
+        { submitting && this.renderSending() }
+
+        { !submitting && complete && failure && this.renderFailure() }
+
+        { !submitting && complete && !failure && this.renderSuccess() }
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js]: I think you could wrap these up into other methods to make it look a bit tidier, e.g.

```jsx
{ this.renderAction() }
{ this.renderStatusMessage() }
```

Then have each method handle its own display logic e.g.

```jsx
renderStatusMessage() {
  const { submitting, complete, failure } = this.state;
  if (submitting) {
     return this.renderSending();
  }
  if (failure) {
    return this.renderFailure();
  }
  return this.renderSuccess();
}
```

File: client/src/components/Icons/CircleTick.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+
+export default ({ color = 'currentColor', size = '3em' }) => (
+  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512" height={size} width={size}>
```
Comment (robbieaverill) [inline client/src/components/Icons/CircleTick.js]: I assume that there are no font icons in the admin module we can use for this?

File: client/src/components/Register.js [hunk 10]
```diff
@@ -226,7 +226,7 @@ class Register extends Component {
 
     // Render loading if we don't have props yet...
     if (!registerProps) {
-      return <LoadingIndicator />;
+      return <LoadingIndicator block color="rgba(0,113,196,1)" />;
```
Comment (robbieaverill) [inline client/src/components/Register.js]: Is there a way we can do this without hard coding colours in? Something with CSS and class names/component modifiers instead?

File: src/Extension/MemberExtension.php [hunk 11]
```diff
@@ -54,7 +54,7 @@ public function updateCMSFields(FieldList $fields)
     {
         $fields->removeByName(['DefaultRegisteredMethodID', 'HasSkippedMFARegistration', 'RegisteredMFAMethods']);
 
-        if (!$this->currentUserCanViewMFAConfig()) {
+        if (!$this->currentUserCanViewMFAConfig() || !$this->owner->exists()) {
```
Comment (robbieaverill) [inline src/Extension/MemberExtension.php]: For a micro-optimisation you could put this check first

File: src/Extension/MemberResetExtension.php [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+     * @return string
+     * @throws PasswordEncryptor_NotFoundException
+     * @throws ValidationException
+     */
+    public function generateAccountResetTokenAndStoreHash(): string
+    {
+        $lifetime = $this->owner->config()->auto_login_token_lifetime;
+
+        do {
+            $generator = new RandomGenerator();
```
Comment (robbieaverill) [inline src/Extension/MemberResetExtension.php]: Do you need to instantiate a new instance for every iteration, or could this line be moved outside the loop?

File: src/Extension/MemberResetExtension.php [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+            $token = $generator->randomToken();
+            $hash = $this->owner->encryptWithUserSettings($token);
+        } while (DataObject::get_one(Member::class, [
+            '"Member"."AccountResetHash"' => $hash,
+        ]));
+
+        $this->owner->AccountResetHash = $hash;
+        $this->owner->AccountResetExpired = date('Y-m-d H:i:s', time() + $lifetime);
```
Comment (robbieaverill) [inline src/Extension/MemberResetExtension.php]:
```suggestion
        $this->owner->AccountResetExpired = DBDatetime::create(DBDatetime::now() + $lifetime)-> Rfc2822();
```

It's a bit more long winded, but follows SS best practices. Related: https://github.com/silverstripe/silverstripe-framework/issues/8802

File: src/Extension/MemberResetExtension.php [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+        return $token;
+    }
+
+    /**
+     * Based on Member::validateAutoLoginToken() and Member::member_from_autologinhash().
+     *
+     * @param string $token
+     * @return Member
+     * @throws PasswordEncryptor_NotFoundException
```
Comment (robbieaverill) [inline src/Extension/MemberResetExtension.php]: This method doesn't throw the exception itself - we usually don't PHPDoc them unless it does

File: src/Extension/MemberResetExtension.php [hunk 15]
```diff
@@ -0,0 +1,81 @@
+<?php
```
Comment (robbieaverill) [inline src/Extension/MemberResetExtension.php]: Can you add strict types?

File: src/Extension/SecurityAdminAccountResetExtension.php [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+                ->addHeader('Content-Type', 'application/json')
+                ->setBody(json_encode(
+                    [
+                        'error' => _t(
+                            __CLASS__ . '.INSUFFICIENT_PERMISSIONS',
+                            'Insufficient permissions to reset user'
+                        )
+                    ]
+                ));
+        }
```
Comment (robbieaverill) [inline src/Extension/SecurityAdminAccountResetExtension.php]: I feel like `jsonResponse()` would be useful here, we've got it in some of the other controllers I think. Would help to reduce some duplicated lines.

File: src/Extension/SecurityAdminAccountResetExtension.php [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+                    'Reset your account'
+                ))
+                ->addData('AccountResetLink', $this->getAccountResetLink($member, $token))
+                ->addData('Member', $member)
+                ->setFrom(Email::config()->admin_email)
+                ->setTo($member->Email);
+
+            return $email->send();
+        } catch (Exception $e) {
+            Injector::inst()->get(LoggerInterface::class)->info('WARNING: Account Reset Email failed to send');
```
Comment (robbieaverill) [inline src/Extension/SecurityAdminAccountResetExtension.php]: Prefer DI configuration here via `private static $dependencies`

File: tests/php/Extension/SecurityAdminAccountResetExtensionTest.php [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+use SilverStripe\Dev\FunctionalTest;
+use SilverStripe\MFA\Extension\SecurityAdminAccountResetExtension;
+use SilverStripe\Security\Member;
+use SilverStripe\Security\SecurityToken;
+
+class SecurityAdminAccountResetExtensionTest extends FunctionalTest
+{
+    protected static $fixture_file = 'SecurityAdminAccountResetExtensionTest.yml';
+
+    public function setUp()
```
Comment (robbieaverill) [inline tests/php/Extension/SecurityAdminAccountResetExtensionTest.php]:
```suggestion
    protected function setUp()
```

File: tests/php/Extension/SecurityAdminAccountResetExtensionTest.php [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        parent::setUp();
+
+        SecurityToken::enable();
+    }
+
+    public function tearDown()
```
Comment (robbieaverill) [inline tests/php/Extension/SecurityAdminAccountResetExtensionTest.php]:
```suggestion
    protected function tearDown()
```

File: client/src/components/Icons/CircleTick.js (hunk 9 above)
Comment (Cheddam) [reply client/src/components/Icons/CircleTick.js]: Only one of the two (the tick) has an equivalent in the admin icon font, and it's not quite the same. For consistency in implementation between the two, I'd like to leave this as-is.

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js (hunk 8 above)
Comment (Cheddam) [reply client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js]: I think we're approaching render method soup on this component (which to be fair probably means I should have child components...), but I get that the grouped checks are kind of hard to parse. I've refactored as requested.

File: client/src/components/Register.js (hunk 10 above)
Comment (Cheddam) [reply client/src/components/Register.js]: Yeah, I wasn't super sure about this implementation, but I also wasn't sure about baking the colours into the icons / CSS, since the front-end app applies a different shade of blue to the admin interface. Maybe there should just be one shade of blue? I'll check with Jared.

File: src/Extension/SecurityAdminAccountResetExtension.php (hunk 16 above)
Comment (Cheddam) [reply src/Extension/SecurityAdminAccountResetExtension.php]: I've written a trait for this method now, can apply it to other classes as we go.

File: src/Extension/MemberResetExtension.php (hunk 12 above)
Comment (Cheddam) [reply src/Extension/MemberResetExtension.php]: This was pulled verbatim from core, so I assumed 'yes', but looking at the implementation of `RandomGenerator` now I don't think it's necessary. I've moved it out.

File: src/Extension/MemberResetExtension.php (hunk 14 above)
Comment (Cheddam) [reply src/Extension/MemberResetExtension.php]: My IDE complains if I don't either add these `@throws` keys or handle the exception within the method 😅 I should probably just turn that warning off, though - it's not really adding value.

File: client/src/components/BasicMath/Login.js [hunk 20]
```diff
@@ -35,7 +35,7 @@ class Login extends Component {
     const { ss: { i18n } } = window;
 
     if (!numbers) {
-      return <LoadingIndicator />;
+      return <LoadingIndicator block color="rgba(0,113,196,1)" />;
```
Comment (robbieaverill) [inline client/src/components/BasicMath/Login.js]: Where'd you get to with these colours?

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+    if (submitting) {
+      return this.renderSending();
+    }
+
+    if (!complete) {
+      return null;
+    }
+
+    return (failed) ? this.renderFailure() : this.renderSuccess();
+  }
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js]: Much tidier, nice work 👍

File: src/Extension/SecurityAdminAccountResetExtension.php [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+            $email = Email::create()
+                ->setHTMLTemplate('SilverStripe\\MFA\\Email\\AccountReset')
+                ->setData($member)
+                ->setSubject(_t(
+                    __CLASS__ . '.ACCOUNT_RESET_EMAIL_SUBJECT',
+                    'Reset your account'
+                ))
+                ->addData('AccountResetLink', $this->getAccountResetLink($member, $token))
+                ->addData('Member', $member)
+                ->setFrom(Email::config()->admin_email)
```
Comment (robbieaverill) [inline src/Extension/SecurityAdminAccountResetExtension.php]:
```suggestion
                ->setFrom(Email::config()->get('admin_email'))
```
Could be personal preference

File: src/Extension/SecurityAdminAccountResetExtension.php [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    protected function sendResetEmail($member)
+    {
+        // Generate / store / obtain reset token
+        $token = $member->generateAccountResetTokenAndStoreHash();
+
+        // Create email and fire
+        try {
+            $email = Email::create()
+                ->setHTMLTemplate('SilverStripe\\MFA\\Email\\AccountReset')
```
Comment (robbieaverill) [inline src/Extension/SecurityAdminAccountResetExtension.php]: We just merged #134 which contains some templates. This looks like a good file path, would you mind updating the others to match?

File: src/Extension/SecurityAdminAccountResetExtension.php [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+                    'Reset your account'
+                ))
+                ->addData('AccountResetLink', $this->getAccountResetLink($member, $token))
+                ->addData('Member', $member)
+                ->setFrom(Email::config()->admin_email)
+                ->setTo($member->Email);
+
+            return $email->send();
+        } catch (Exception $e) {
+            $this->logger->info('WARNING: Account Reset Email failed to send');
```
Comment (robbieaverill) [inline src/Extension/SecurityAdminAccountResetExtension.php]: Want to include the exception message?

File: client/src/components/Register.js (hunk 10 above)
Comment (Cheddam) [reply client/src/components/Register.js]: Jared was happy to unify the colours, so this is now baked into the component.

File: client/src/components/BasicMath/Login.js (hunk 20 above)
Comment (Cheddam) [reply client/src/components/BasicMath/Login.js]: Just pushed a resolution 👍

File: templates/SilverStripe/MFA/Email/AccountReset.ss [hunk 25]
```diff
@@ -0,0 +1,5 @@
+<p><%t SilverStripe\\Control\\ChangePasswordEmail_ss.HELLO 'Hi' %> $Member.FirstName,</p>
```
Comment (robbieaverill) [inline templates/SilverStripe/MFA/Email/AccountReset.ss]: I think you should include the name as part of the translation e.g. `"Hi {name}"` - not sure about the comma, perhaps also in the translation

Comment (Cheddam) [reply templates/SilverStripe/MFA/Email/AccountReset.ss]: There's no variation of the core translation that includes the name by the looks of things - I'm inclined to leave this as-is for maximum localisation.

File: src/Extension/SecurityAdminAccountResetExtension.php (hunk 23 above)
Comment (Cheddam) [reply src/Extension/SecurityAdminAccountResetExtension.php]: Done!
